### PR TITLE
[clang] move -Wcast-function-type under -Wextra

### LINKED ISSUE #76872: [clang] move -Wcast-function-type under -Wextra
via https://github.com/llvm/llvm-project/issues/74617, I was looking into why I observe this diagnostic with gcc but not clang. It looks like our project (llvmlibc) builds with -Wextra but not -Wcast-function-type.   GCC 8 added support for that flag, and did so under -Wextra.

### PR DESCRIPTION
**Pull request by Abhinkop**
The -Wcast-fuction-type-strict has been moved under dignstic group -Wextra.
Edited the test cases for -Wcast-fuction-type-strict and -Wcast-fuction-type in Sema an SemaCXX.

Added a new test case which include a functionality that was already in the -Wextra group, i.e -Wignored-qualifiers with -Wcast-fuction-type-strict.

Fixes: #76872

### CODE REVIEW

**Comment by Abhinkop** [comment]
@nickdesaulniers

**Review by shafik — COMMENTED** [review]
This should have a release note and I think this is a potentially breaking change since folks using `Wextra` may get this diagnostic now.

File: clang/test/Sema/warn-extra-cast-function-type-strict.c [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// RUN: %clang_cc1 %s -fsyntax-only -Wextra -verify
+
+
+int t(int array[static 12]);
+int u(int i);
+const int v(int i); /* expected-warning {{'const' type qualifier on return type has no effec}} */
```
Comment (Endilll) [inline clang/test/Sema/warn-extra-cast-function-type-strict.c:6]:
```suggestion
const int v(int i); /* expected-warning {{'const' type qualifier on return type has no effect}} */
```

File: clang/test/Sema/warn-extra-cast-function-type-strict.c [hunk 2]
```diff
@@ -0,0 +1,42 @@
+// RUN: %clang_cc1 %s -fsyntax-only -Wextra -verify
```
Comment (Endilll) [inline clang/test/Sema/warn-extra-cast-function-type-strict.c:1]: As written, tests checks only default language mode for C (which is `gnu99` AFAIK). Maybe we want to check other modes as well?

File: clang/test/Sema/warn-extra-cast-function-type-strict.c [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  a = (f1 *)x;
+  b = (f2 *)x; /* expected-warning {{cast from 'int (*)(long)' to 'f2 *' (aka 'int (*)(void *)') converts to incompatible function type}} */
+  c = (f3 *)x; /* expected-warning {{cast from 'int (*)(long)' to 'f3 *' (aka 'int (*)()') converts to incompatible function type}} */
+  d = (f4 *)x; /* expected-warning {{cast from 'int (*)(long)' to 'f4 *' (aka 'void (*)()') converts to incompatible function type}} */
+  e = (f5 *)x; /* expected-warning {{cast from 'int (*)(long)' to 'f5 *' (aka 'void (*)(void)') converts to incompatible function type}} */
+  f = (f6 *)x; /* expected-warning {{cast from 'int (*)(long)' to 'f6 *' (aka 'int (*)(long, int)') converts to incompatible function type}} */
+  g = (f7 *)x; /* expected-warning {{cast from 'int (*)(long)' to 'f7 *' (aka 'int (*)(long, ...)') converts to incompatible function type}} */
+  h = (f8 *)t;
+  i = (f9 *)u;
+  j = (f10 *)v; /* expected-warning {{cast from 'const int (*)(int)' to 'f10 *' (aka 'int (*)(int)') converts to incompatible function type}} */
```
Comment (Endilll) [inline clang/test/Sema/warn-extra-cast-function-type-strict.c:41]: On line 6 we issue a diagnostic that `const` qualifier doesn't have effect on return type, yet we issue a diagnostic here telling that this makes function pointer type incompatible. It feels like we fail to make up our minds on the matter, or that we are overly pedantic without pedantic warnings enabled.

**Review by Endilll — COMMENTED** [review]
I think this PR moves things in the right direction, but I'm not well-versed in C enough to approve.

> This should have a release note and I think this is a potentially breaking change since folks using Wextra may get this diagnostic now.

I agree that release note should be added, but I'm not sure we break people with this. We ship new versions with new on-by-default diagnostics. How is this different? Not to say that this improves GCC compatibility (more on this below).

Out of 7 warning we expect in the test, GCC 13.1 issues a warning only for `f2`, `f4`, and `f6` cases: https://godbolt.org/z/18Eqnsqba. Same goes for GCC trunk. I'm not qualified enough in C to be sure which compiler gets this right. @AaronBallman would be very welcome here.

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 2 above)
Comment (Abhinkop) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:1]: add runs for -std=cNN for all remaining 5 of them? and also add extra test cases where conflicts arise?

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 3 above)
Comment (Abhinkop) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:41]: Yeah, even I found this strange. Is there a place we can verify which is the correct behavior?

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 2 above)
Comment (Endilll) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:1]: Yes, more `RUN` lines with different `-std=` options, and potentially more `expected` directives tailored for particular language mode. We have an example of this in C++ defect report tests: https://github.com/llvm/llvm-project/blob/main/clang/test/CXX/drs/dr6xx.cpp

Note that I might be biased here, as someone who've spent a significant time with C++ defect report test suite.

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 3 above)
Comment (Endilll) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:41]: I hope @AaronBallman can clarify.

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 2 above)
Comment (nickdesaulniers) [inline clang/test/Sema/warn-extra-cast-function-type-strict.c:1]: If this diagnostic is not already well tested, adding all of these additional unit tests makes sense.

But it looks like there's quite a number of existing tests in other files.  I don't think we should add a new file just to check this diagnostics behavior via `-Wextra` via a whole new file.

**Comment by nickdesaulniers** [comment]
Github trick not specific to llvm-project.  If you put

```
Fixes: #76872
```
In the commit message, then merging this PR will automatically close https://github.com/llvm/llvm-project/issues/76872.

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 2 above)
Comment (Abhinkop) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:1]: No I was thinking of adding extra runs for both -Wcast-function-type and -Wextra or should the one for -Wcast-function-type be done in a different issue?

Comment (nickdesaulniers) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:1]: I would just remove the newly created clang/test/Sema/warn-extra-cast-function-type-strict.c.  If you have cases that expose bugs in the implementation, that is distinctly a separate issue from adding this warning under -Wextra.

Comment (Abhinkop) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:1]: Ok. sounds good I will do that.

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 3 above)
Comment (AaronBallman) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:41]: I think the diagnostic here is incorrect.

C23 6.7.6.3p4:
If, in the declaration "T D1", D1 has the form
  D ( parameter-type-listopt ) attribute-specifier-sequenceopt
and the type specified for ident in the declaration "T D" is "derived-declarator-type-list T", then the
type specified for ident is "derived-declarator-type-list function returning the unqualified, non-atomic
version of T". The optional attribute specifier sequence appertains to the function type.

So the type of the function does not include the qualifiers on the return type, which means that `const int f(void)` and `int f(void)` should result in the same type.

We get this wrong: https://godbolt.org/z/xvh449xrd

I think this is a more general problem with how we model the function type: https://godbolt.org/z/G3vxjW9dh

Comment (Abhinkop) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:41]: So, create an issue? Comment out this test case and put a comment relating to the new issue?

Comment (AaronBallman) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:41]: No need to create an issue, we've already got one: https://github.com/llvm/llvm-project/issues/39494

Comment (AaronBallman) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:41]: Probably worth it to add a FIXME comment above the test case and point to that issue, though.

Comment (Abhinkop) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:41]: So finally the things. I need to do:
Add FIXME comment and leave the test case as is. Remove the clang/test/Sema/warn-extra-cast-function-type-strict.c file . Update release notes. Then update the pull request. Agree? @nickdesaulniers @AaronBallman @Endilll @shafik

Comment (AaronBallman) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c:41]: That sounds correct to me

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 1 above)
Comment (Abhinkop) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c]: Didn't have to do it as the file is now not added.

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 3 above)
Comment (Abhinkop) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c]: Done.

File: clang/test/Sema/warn-extra-cast-function-type-strict.c (hunk 2 above)
Comment (Abhinkop) [reply clang/test/Sema/warn-extra-cast-function-type-strict.c]: Done. File removed, and a FIXME comment added for the existing test.

**Comment by Abhinkop** [comment]
@nickdesaulniers @AaronBallman Can you'll take a look at this pull request?

File: clang/docs/ReleaseNotes.rst [hunk 4]
```diff
@@ -164,6 +164,8 @@ Deprecated Compiler Flags
 Modified Compiler Flags
 -----------------------
 
+* ``-Wextra`` group flag now contains an additional flag i.e, ``-Wcast-function-type``.
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:167]:
```suggestion
- Added ``-Wcast-function-type`` as a warning enabled by ``-Wextra``. #GH76872
```

**Review by AaronBallman — APPROVED** [review]
LGTM! I'll fix up the nit and land on your behalf; thank you for the improvement!

**Comment by amy-kwan** [comment]
I believe this patch is causing some issues on two PPC bots. Would you be able to help take a look?
https://lab.llvm.org/buildbot/#/builders/57/builds/33601/steps/5/logs/stdio
https://lab.llvm.org/buildbot/#/builders/36/builds/43759/steps/12/logs/stdio

**Comment by Abhinkop** [comment]
> I believe this patch is causing some issues on two PPC bots. Would you be able to help take a look? https://lab.llvm.org/buildbot/#/builders/57/builds/33601/steps/5/logs/stdio https://lab.llvm.org/buildbot/#/builders/36/builds/43759/steps/12/logs/stdio

I guess adding this flag (-Wno-cast-function-type-strict) will make the error go away. But it is caused by the conversion of the second argument "void*" to "siginfo_t *" when casting from "SignalHandlerType" i.e. "void (*__sanitizer::SignalHandlerType)(int, void *, void *);" to "void (*sa_sigaction) (int, siginfo_t *, void *);". If you are sure that this conversion is valid, I guess we can add this file. Might be similar for others.

**Comment by AaronBallman** [comment]
I'm confused as to how this code ever compiled in the first place... In each case, this is C++ code that's failing:
```
/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:219:25: error: cast from 'SignalHandlerType' (aka 'void (*)(int, void *, void *)') to 'sa_sigaction_t' (aka 'void (*)(int, siginfo_t *, void *)') converts to incompatible function type [-Werror,-Wcast-function-type-strict]

/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:437:4: error: cast from 'void (*)()' to 'void (*)(void *)' converts to incompatible function type [-Werror,-Wcast-function-type-strict]

/home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:460:54: error: cast from 'void (*)(void *)' to 'void (*)()' converts to incompatible function type [-Werror,-Wcast-function-type-strict]
```
All of which is being compiled in `-std=c++17` and none of which are valid in C++ and the behavior is the same between Clang 18, trunk, and GCC: https://godbolt.org/z/WY9zvsa8z

@amy-kwan we may need some help from you with investigating this; but in the meantime, this commit can be reverted to get the bots back to green if that's blocking you.

**Comment by Abhinkop** [comment]
> I'm confused as to how this code ever compiled in the first place... In each case, this is C++ code that's failing:
> 
> ```
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:219:25: error: cast from 'SignalHandlerType' (aka 'void (*)(int, void *, void *)') to 'sa_sigaction_t' (aka 'void (*)(int, siginfo_t *, void *)') converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> 
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:437:4: error: cast from 'void (*)()' to 'void (*)(void *)' converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> 
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:460:54: error: cast from 'void (*)(void *)' to 'void (*)()' converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> ```
> 
> All of which is being compiled in `-std=c++17` and none of which are valid in C++ and the behavior is the same between Clang 18, trunk, and GCC: https://godbolt.org/z/WY9zvsa8z
> 
> @amy-kwan we may need some help from you with investigating this; but in the meantime, this commit can be reverted to get the bots back to green if that's blocking you.

@AaronBallman There was a forced/explicit typecasting done, and hence the compiler didn't complain. for example, https://godbolt.org/z/G94r7qvxh . When the -Werror flag coupled with -Wextra flag caught this conversion, for example, https://godbolt.org/z/6bMrdfe4n. I hope this might help to provide some context.

**Comment by amy-kwan** [comment]
> I'm confused as to how this code ever compiled in the first place... In each case, this is C++ code that's failing:
> 
> ```
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:219:25: error: cast from 'SignalHandlerType' (aka 'void (*)(int, void *, void *)') to 'sa_sigaction_t' (aka 'void (*)(int, siginfo_t *, void *)') converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> 
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:437:4: error: cast from 'void (*)()' to 'void (*)(void *)' converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> 
> /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:460:54: error: cast from 'void (*)(void *)' to 'void (*)()' converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> ```
> 
> All of which is being compiled in `-std=c++17` and none of which are valid in C++ and the behavior is the same between Clang 18, trunk, and GCC: https://godbolt.org/z/WY9zvsa8z
> 
> @amy-kwan we may need some help from you with investigating this; but in the meantime, this commit can be reverted to get the bots back to green if that's blocking you.

@AaronBallman Sounds good. I am ok with this commit being reverted and we can investigate on the side.

**Comment by AaronBallman** [comment]
> > I'm confused as to how this code ever compiled in the first place... In each case, this is C++ code that's failing:
> > ```
> > /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:219:25: error: cast from 'SignalHandlerType' (aka 'void (*)(int, void *, void *)') to 'sa_sigaction_t' (aka 'void (*)(int, siginfo_t *, void *)') converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> > 
> > /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:437:4: error: cast from 'void (*)()' to 'void (*)(void *)' converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> > 
> > /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:460:54: error: cast from 'void (*)(void *)' to 'void (*)()' converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > All of which is being compiled in `-std=c++17` and none of which are valid in C++ and the behavior is the same between Clang 18, trunk, and GCC: https://godbolt.org/z/WY9zvsa8z
> > @amy-kwan we may need some help from you with investigating this; but in the meantime, this commit can be reverted to get the bots back to green if that's blocking you.
> 
> @AaronBallman There was a forced/explicit typecasting done, and hence the compiler didn't complain. for example, https://godbolt.org/z/G94r7qvxh . When the -Werror flag coupled with -Wextra flag caught this conversion, for example, https://godbolt.org/z/6bMrdfe4n. I hope this might help to provide some context.

Ah! That does provide the context I needed, thank you! And I think I see the issue:

https://github.com/llvm/llvm-project/blob/734026347cca85cf0e242ef5f04896f55e0ac113/clang/include/clang/Basic/DiagnosticGroups.td#L576

`-Wcast-function-type` also includes `-Wcast-function-type-strict`; we don't want the strict warning in `-Wextra`, just the base warning. I'm looking to see if there's a reasonable way to accomplish this.

**Comment by Abhinkop** [comment]
> > > I'm confused as to how this code ever compiled in the first place... In each case, this is C++ code that's failing:
> > > ```
> > > /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:219:25: error: cast from 'SignalHandlerType' (aka 'void (*)(int, void *, void *)') to 'sa_sigaction_t' (aka 'void (*)(int, siginfo_t *, void *)') converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> > > 
> > > /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:437:4: error: cast from 'void (*)()' to 'void (*)(void *)' converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> > > 
> > > /home/buildbots/llvm-external-buildbots/workers/ppc64le-clang-rhel-test/clang-ppc64le-rhel/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:460:54: error: cast from 'void (*)(void *)' to 'void (*)()' converts to incompatible function type [-Werror,-Wcast-function-type-strict]
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > All of which is being compiled in `-std=c++17` and none of which are valid in C++ and the behavior is the same between Clang 18, trunk, and GCC: https://godbolt.org/z/WY9zvsa8z
> > > @amy-kwan we may need some help from you with investigating this; but in the meantime, this commit can be reverted to get the bots back to green if that's blocking you.
> > 
> > 
> > @AaronBallman There was a forced/explicit typecasting done, and hence the compiler didn't complain. for example, https://godbolt.org/z/G94r7qvxh . When the -Werror flag coupled with -Wextra flag caught this conversion, for example, https://godbolt.org/z/6bMrdfe4n. I hope this might help to provide some context.
> 
> Ah! That does provide the context I needed, thank you! And I think I see the issue:
> 
> https://github.com/llvm/llvm-project/blob/734026347cca85cf0e242ef5f04896f55e0ac113/clang/include/clang/Basic/DiagnosticGroups.td#L576
> 
> `-Wcast-function-type` also includes `-Wcast-function-type-strict`; we don't want the strict warning in `-Wextra`, just the base warning. I'm looking to see if there's a reasonable way to accomplish this.

Change def CastFunctionType : DiagGroup<"cast-function-type", [CastFunctionTypeStrict]>; to def  CastFunctionType : DiagGroup<"cast-function-type">; and introduce a new one def CastFunctionTypeDefault : DiagGroup<[CastFunctionType], [CastFunctionTypeStrict]>; and some how pass this on where CastFunctionType was used earlier?

**Comment by AaronBallman** [comment]
Yeah, I was hoping to avoid having to introduce another diagnostic group, but I think that's going to be the best way forward. I've got a patch in progress to add `-Wcast-function-type-mismatch` so we end up with:
```
-Wcast-function-type
|- -Wcast-function-type-mismatch
|- -Wcast-function-type-strict
```
and then add `-Wcast-function-type-mismatch` to `-Wextra`, but still thinking through the changes and whether there's a different approach.

**Comment by AaronBallman** [comment]
I've put up a PR for review at https://github.com/llvm/llvm-project/pull/86131

**Comment by thurstond** [comment]
> > I believe this patch is causing some issues on two PPC bots. Would you be able to help take a look? https://lab.llvm.org/buildbot/#/builders/57/builds/33601/steps/5/logs/stdio https://lab.llvm.org/buildbot/#/builders/36/builds/43759/steps/12/logs/stdio
> 
> I guess adding this flag (-Wno-cast-function-type-strict) will make the error go away. But it is caused by the conversion of the second argument "void*" to "siginfo_t _" when casting from "SignalHandlerType" i.e. "void (___sanitizer::SignalHandlerType)(int, void *, void *);" to "void (*sa_sigaction) (int, siginfo_t *, void *);". If you are sure that this conversion is valid, I guess we can add this flag. Might be similar for others.

The reason for the weird cast: SignalHandlerType has 'void*' instead of 'siginfo_t*' because it is typedef'ed in sanitizer_common/sanitizer_common.h, which does not have access to the header (signal.h) that defines siginfo_t; we therefore cannot fix SignalHandlerType.

**Pull request merged**